services/src/main/java/org/keycloak/services/filters/CachingGZIPEncodingInterceptor.java

```diff
@@ -0,0 +1,90 @@
+package org.keycloak.services.filters;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.output.TeeOutputStream;
+import org.jboss.resteasy.plugins.interceptors.encoding.GZIPEncodingInterceptor;
+import org.jboss.resteasy.resteasy_jaxrs.i18n.LogMessages;
+import org.keycloak.common.Profile;
+
+import javax.annotation.Priority;
+import javax.ws.rs.Priorities;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.UriInfo;
+import javax.ws.rs.ext.Provider;
+import javax.ws.rs.ext.WriterInterceptorContext;
+import java.io.ByteArrayOutputStream;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+@Provider
+@Priority(Priorities.ENTITY_CODER)
+public class CachingGZIPEncodingInterceptor extends GZIPEncodingInterceptor {
+
+    private final Path tmpDir;
+    @Context
+    private UriInfo uriInfo;
+
+    public CachingGZIPEncodingInterceptor() {
+        try {
+            this.tmpDir = Files.createTempDirectory("gzip-cache");
+        } catch (IOException e) {
+            throw new RuntimeException("could not create temp gzip cache directory");
+        }
+    }
+
+    public static class CacheCommittedGZIPOutputStream extends CommittedGZIPOutputStream {
+        protected CacheCommittedGZIPOutputStream(final OutputStream delegate) {
+            super(delegate, null);
+        }
+    }
+
+    @Override
+    public void aroundWriteTo(WriterInterceptorContext context) throws IOException, WebApplicationException {
+        LogMessages.LOGGER.debugf("Interceptor : %s,  Method : aroundWriteTo", getClass().getName());
+
+        Object encoding = context.getHeaders().getFirst(HttpHeaders.CONTENT_ENCODING);
+        final boolean featureEnabled = Profile.isFeatureEnabled(Profile.Feature.GZIP);
+
+        if (encoding != null && encoding.toString().equalsIgnoreCase("gzip") && featureEnabled) {
+            OutputStream old = context.getOutputStream();
+
+            final String fileName = uriInfo.getPath();
+            final Path zipFile = Paths.get(tmpDir.toString(), fileName);
+
+            // Any content length set will be obsolete
+            context.getHeaders().remove("Content-Length");
```

Comment:
Surely content-length should still be sent?

Reply:
a wrong one?

Reply:
If compressed data is cached we'd have the real content-length wouldn't we?

Reply:
Simple solution could be to just created the cache, then return it afterwards, rather than doing the tee stream, as then you'd have the correct content-length.

Reply:
There is really no need to set the right `Content-Length` as we use `Transfer-Encoding: chunked`

Reply:
Seems to me that it would be better to set the content-length, since we actually know it. From what I could tell on Google, Twitter, etc. They all send content-length for compressed files.

Reply:
Well we don't know it unless we do the compression in memory write to disk and to the browser. With the current implementation we write to disk and to the browser simultaneously and compress on the fly. Because we use chunked encoding the receiver can still tell if it received all the content and will not cache something partial.

[original rest-easy implementation](https://java-browser.yawk.at/org.jboss.resteasy/resteasy-jaxrs/3.6.1.Final/org/jboss/resteasy/plugins/interceptors/encoding/GZIPEncodingInterceptor.java#87) also removes the `Content-Lenght`